### DISCUSSION TITLE
Time conversion from 'time_signature' and BPM

### DISCUSSION THREAD

**Discussion by joel-tuberosa** [mido/mido#347] [Q&A]
Hello,

I have exported a MIDI file from Ableton Live 11 and I would like to extract the variation of the mod wheel in the form of a two-columns table with values and corresponding timestamp in seconds.

For this I have written a Python script using mido to parse the MIDI file. As in this file, there is no 'set_tempo' MetaMessage I provided the BPM value as a parameter to my script. Using the 'time_signature' MetaMessage, I was thinking to convert MIDI time into seconds using the clocks_per_click value. If I got it right, clocks_per_click is the number of MIDI clock ticks per quarter of note. Therefore, my MIDI time to second conversion should be:

```
# number of seconds in a beat
beat_seconds = 60 / bpm

# number of seconds in a quarter note
quarter_note_seconds = beat_seconds / 4 

# number of seconds in a MIDI clock tick
tick_seconds = quarter_note_seconds / clocks_per_click

# MIDI time to second conversion
seconds = time * tick_seconds
```

The issue is that the MIDI file from Ableton display 36 clocks_per_click, though it doesn't give me the correct result. The correct result is obtained when clocks_per_click is set to 24.

Here is the MIDI file I am referring to: [modramp.zip](https://github.com/mido/mido/files/7579290/modramp.zip) 
It is a linear increase of the mod wheel from 0 to 255 during the first beat. The MIDI last one measure.

I would really appreciate if you can explain me why the clocks_per_click seems not to correspond to the time displayed in the messages there and if there would be another way to convert this time value without using the 'set_tempo' MetaMessage, which is missing.

Thank you in advance,

Joel

**Comment by rdoursenaud** [comment]
You may want to have a look at https://mido.readthedocs.io/en/1.2.10/midi_files.html?highlight=time#about-the-time-attribute and check https://github.com/mido/mido/pull/114

**Comment by intcreator** [comment]
According to the [Ableton Live docs ](https://help.ableton.com/hc/en-us/articles/209068169-Understanding-MIDI-files#exportresolution) the MIDI export resolution is **96 ppq**.

Using Mido I can convert ticks to seconds using the following:

```python
import mido

ticks_per_beat = 96
bpm = 118

ticks = 2688

tempo = mido.bpm2tempo(bpm)
seconds = mido.tick2second(ticks, ticks_per_beat, tempo)
```

Aside for OP's specific case: It says "which means a 16th note can be divided into 24 steps" so that might be where your 24 number came in. With common time (4/4) a quarter note is the same as one beat, so your `quarter_note_seconds` should be the same as `beat_seconds`. Dividing it by 4 would give you `sixteenth_note_seconds`.

Not sure why the Live export has that 36 in there though. That confused me too.
